device/transport/amqp/src/amqp.ts

```diff
@@ -297,15 +303,17 @@ export class Amqp extends EventEmitter implements Client.Transport {
               });
             }
           },
-          disconnect: (disconnectCallback) => this._fsm.transition('disconnecting', disconnectCallback),
+          disconnect: (disconnectCallback) => this._fsm.transition('disconnecting', null, disconnectCallback),
+          /*Codes_SRS_NODE_DEVICE_AMQP_16_081: [if the handler specified in the `setDisconnectHandler` call is called while the `Amqp` object is connecting or authenticating, the connection shall be stopped and an `disconnect` event shall be emitted with the error translated to a transport-agnostic error.]*/
+          disconnectError: (err, callback) => this._fsm.transition('disconnecting', err, callback),
```

Comment:
>disconnectError [](start = 10, length = 15)

I have a question here that comes back to the name of this action.  If this action happens because we've already been disconnected, is "disconnecting" the right state to go into?  If the network connection has already been severed, then we just want to clean up our internal housekeeping, but I don't think we want to call into amqp10 to do anything.  According to my guess, amqp already considers this disconnected, so why  risk problems by calling functions based on the assumption that parts of our connection might still be open?

Reply:
yes (and that's what happens in the underlying amqp object). I guess the "disconnecting" state could be renamed to "cleaning up". if you look at the code in the `disconnecting._onEnter` you'll see that there are 2 paths depending if there's an error or not: one of them cleanly shuts down resources, the other one force-detaches everything (which doesn't send anything on the wire, it just resets state machines).

---
In reply to: [175764457](https://github.com/Azure/azure-iot-sdk-node/pull/254#discussion_r175764457) [](ancestors = 175764457)